service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +461,233 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+      snapshotId = snap.parentId();
+      // we need continuous sequence number to correctly rollback
+      expectedSequenceNumber--;
+    }
+
+    boolean wasExpectedSnapshotReached = Objects.equals(snapshotId, expectedCurrentSnapshotId);
+    updateToRemoveSnapshot.add(new MetadataUpdate.RemoveSnapshots(snapshotsToRemove));
+    return wasExpectedSnapshotReached ? updateToRemoveSnapshot : null;
+  }
+
+  private boolean isRollbackSnapshot(Snapshot snapshot) {
+    // Only Snapshots with {@ROLLBACKABLE_REPLACE_SNAPSHOT} are allowed to be rollback.
+    return DataOperations.REPLACE.equals(snapshot.operation())
+        && PropertyUtil.propertyAsString(snapshot.summary(), CONFLICT_RESOLUTION_ACTION, "")
+            .equalsIgnoreCase("rollback");
+  }
+
+  private MetadataUpdate.SetSnapshotRef findSetSnapshotRefUpdate(UpdateTableRequest request) {
+    int total = 0;
+    MetadataUpdate.SetSnapshotRef setSnapshotRefUpdate = null;
+    // find the SetRefName snapshot update
+    for (MetadataUpdate update : request.updates()) {
+      if (update instanceof MetadataUpdate.SetSnapshotRef) {
+        total++;
+        setSnapshotRefUpdate = (MetadataUpdate.SetSnapshotRef) update;
+      }
+    }
+
+    // if > 1 assertion for refs, then it's not safe to rollback, make this Noop.
+    return total != 1 ? null : setSnapshotRefUpdate;
+  }
+
+  private MetadataUpdate.AddSnapshot findAddSnapshotUpdate(UpdateTableRequest request) {
+    int total = 0;
+    MetadataUpdate.AddSnapshot addSnapshot = null;
+    // find the SetRefName snapshot update
+    for (MetadataUpdate update : request.updates()) {
+      if (update instanceof MetadataUpdate.AddSnapshot) {
+        total++;
+        addSnapshot = (MetadataUpdate.AddSnapshot) update;
+      }
+    }
+
+    // if > 1 assertion for addSnapshot, then it's not safe to rollback, make this Noop.
+    return total != 1 ? null : addSnapshot;
+  }
+
+  private TableMetadata setAppropriateLastSeqNumber(TableMetadata newBase) {
+    // TODO: Get rid of the reflection call once TableMetadata have API for it.
+    // move the lastSequenceNumber back, to apply snapshot properly on the
+    // current-metadata Seq number are considered increasing monotonically
+    // snapshot over snapshot, the client generates the manifest list and hence
+    // the sequence number can't be changed for a snapshot the only possible option
+    // then is to change the sequenceNumber tracked by metadata.json
+    try {
+      long lastSeqNumber = newBase.lastSequenceNumber();
+      // this should point to the sequence number that current tip of the
+      // branch belongs to, as the new commit will be applied on top of this.
+      LAST_SEQUENCE_NUMBER_FIELD.set(newBase, newBase.currentSnapshot().sequenceNumber());
+      LOGGER.info(
+          "Setting table :{} last sequence number from {} to {}",
+          newBase.uuid(),
+          lastSeqNumber,
+          newBase.lastSequenceNumber());
```

Comment:
for humans to read this we probably need the table's name too? WDYT?

Reply:
I think table UUID is not easily usable by humans for trouble-shooting, but I do not insist.